transactions/multi-document-acid-transactions-mongodb-go.md

```diff
@@ -0,0 +1,214 @@
+
+Take the following schema validation logic:
+
+```json
+{
+    "$jsonSchema": {
+        "additionalProperties": true,
+        "properties": {
+            "duration": {
+                "bsonType": "int",
+                "minimum": 2
+            }
+        }
+    }
+}
+```
+
+The above logic would be applied using the MongoDB CLI or with Compass, but we're essentially saying that our schema for the `episodes` collection can contain any fields in a document, but the `duration` field must be an integer and it must be at least two. Could our schema validation be more complex? Absolutely, but we're all about simplicity in this example. If you want to learn more about schema validation, check out [this awesome tutorial](https://www.mongodb.com/blog/post/json-schema-validation--locking-down-your-model-the-smart-way) on the subject.
+
+Now that we know the schema and what will cause a failure, we can start implementing some transaction code that will commit or roll back changes.
+
+## Starting and Committing Transactions
+
+Before we dive into starting a session for our operations and committing transactions, let's establish a base point in our project. Let's assume that your project has the following boilerplate MongoDB with Go code:
+
+```go
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Episode represents the schema for the "Episodes" collection
+type Episode struct {
+	ID          primitive.ObjectID `bson:"_id,omitempty"`
+	Podcast     primitive.ObjectID `bson:"podcast,omitempty"`
+	Title       string             `bson:"title,omitempty"`
+	Description string             `bson:"description,omitempty"`
+	Duration    int32              `bson:"duration,omitempty"`
+}
+
+func main() {
+	client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(os.Getenv("ATLAS_URI")))
+	if err != nil {
+		panic(err)
+	}
+	defer client.Disconnect(context.TODO())
+
+	database := client.Database("quickstart")
+	episodesCollection := database.Collection("episodes")
+}
+```
+
+Now let's also assume that you've correctly included the MongoDB Go driver as seen in a previous tutorial titled, [How to Get Connected to Your MongoDB Cluster with Go](https://www.mongodb.com/blog/post/quick-start-golang--mongodb--starting-and-setup).
+
+The goal here will be to try to insert a document that complies with our schema validation as well as a document that doesn't, so that we have a commit that doesn't happen.
+
+```go
+// ...
+
+func main() {
+    // ...
+
+	session, err := client.StartSession()
+	if err != nil {
+		panic(err)
+	}
+	defer session.EndSession(context.Background())
+
+	err = mongo.WithSession(context.Background(), session, func(sessionContext mongo.SessionContext) error {
```

Comment:
I incorrectly resolved a comment thread from before because I didn't think we were still showing an example of `WithSession`, but this does need an explicit `AbortTransaction` call if the txn fails. Easiest way to do this would probably outside the `WithSession` call if the error returned is non-nil.

Also, all of the `panic(err)` calls inside `WithSession` should be changed to `return err` so the error is returned and processed correctly outside of the callback.